src/io.rs

```diff
@@ -0,0 +1,95 @@
+use std::io::{Error, Read, Result, Write};
+
+use libc::{c_void, size_t};
+
+use crate::error::rustls_io_error;
+
+/// A callback for rustls_server_session_read_tls or rustls_client_session_read_tls.
+/// An implementation of this callback should attempt to read up to n bytes from the
+/// network, storing them in `buf`. If any bytes were stored, the implementation should
+/// set out_n to the number of bytes stored and return 0. If there was an error,
+/// (including EAGAIN or EWOULDBLOCK), the implementation should return `errno`.
+/// It's best to make one read attempt to the network per call. Additional reads will
+/// be triggered by subsequent calls to one of the `_read_tls` methods.
+/// `userdata` is set to the value provided to `rustls_*_session_set_userdata`. In most
+/// cases that should be a struct that contains, at a minimum, a file descriptor.
+/// The buf and out_n pointers are borrowed and should not be retained across calls.
+pub type rustls_read_callback = Option<
+    unsafe extern "C" fn(
+        userdata: *mut c_void,
+        buf: *mut u8,
+        n: size_t,
+        out_n: *mut usize,
```

Comment:
Can `out_n` be `*mut size_t`, since it and `n` both represent a count of bytes? I also think the C representation is unfortunate, since the `usize` ends up as `uintptr_t` and to me, `uintptr_t` is a sign that you're doing something suspicious.